DOCS-8181-4.3-tuning-monitoring-kt (#1437)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-monitoring.adoc

> = Performance Monitoring

Reviewer: copyedit only

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> To complement the best practices for performance, it is necessary to monitor the exhausted and consumed resources by Mule runtime engine (Mule) during the tests. The following categories can help to monitor the main system resources. Additionally, learn what tools to use for CloudHub monitoring.  +

Reviewer: Do you mean "to complement the best practices for performance TESTING"?
```suggestion
To complement performance testing best practices, monitor the resources consumed by Mule runtime engine (Mule) during the tests. The following information identifies the metrics that are most important to monitor for each main system resource and the tools to use for CloudHub monitoring. +
```

> The commands provided in the next sections are for a Linux SO with sysstat installed.

Reviewer: Do you mean "Linux OS"?
```suggestion
The resource monitoring commands documented here are for a Linux OS with sysstat installed.
```
What if the audience is not using Linux with sysstat?

== CPU Resource

> Increased load can exhaust CPU resources. There are several ways to monitor these resources, two of them are:

Reviewer: ```suggestion
Increased load can exhaust CPU resources. There are several ways to monitor these resources. Two of them are:
```

* Top command: `top` +
The `%CPU` metric equals to the CPU usage divided by the number of cores.

.Top command CPU stats
image::mruntime-tuning-monitoring-top.png[Top command CPU stats]

* Sar command: `sar -u  ALL $INTERVAL $COUNT` +

> Where `$INTERVAL` specifies the time interval to obtain the metrics, and `$COUNT` specifies the number of data points to collect. +

Reviewer: ```suggestion
`$INTERVAL` specifies the time interval allocated for obtaining the metrics and `$COUNT` specifies the number of data points to collect. +
```

> Main metrics to pay attention are:

Reviewer: ```suggestion
The main metrics to pay attention to are:
```

> ** `%usr` shows how much CPU is used at the user level (application).

Reviewer: ```suggestion
** `%usr`, which shows how much CPU is used at the user level (application)
```

> ** `%sys` shows how much CPU is used at system level (kernel).

Reviewer: ```suggestion
** `%sys`, which shows how much CPU is used at system level (kernel)
```

> ** `%idle` shows the percentage of time that the CPU was idle.

Reviewer: ```suggestion
** `%idle`, which shows the percentage of time that the CPU was idle.
```

.Sar command CPU stats
image::mruntime-tuning-sar-cpu.png[Sar command CPU stats]

== Memory Resource

> To understand how Mule uses memory, including heap and native memory or off-heap buffers, monitor the memory via several ways, two of them are:

Reviewer: ```suggestion
To understand how Mule uses memory, including heap and native memory or off-heap buffers, monitor the memory  in any of several ways. Two of them are:
```

* GC Log +

> Enable GC (Garbage Collection) logging to obtain allocation rate, object growth, and so on. Additionally, obtain information about the activities in the heap. Notice that enabling GC logging has a light overhead on the disk.

Reviewer: ```suggestion
Enable GC (Garbage Collection) logging to obtain allocation rate, object growth, and so on. Additionally, obtain information about the activities in the heap. Enabling GC logging involves a light disk overhead.
```

.CG Log wrapper.conf
image::mruntime-tuning-wrapper.png[wrapper.conf]

* Sar command: `sar -r $INTERVAL $COUNT` +

> Main metrics to pay attention are:

Reviewer: ```suggestion
The main metrics to pay attention to are:
```
Search for this text and make this edit GLOBALLY; I will stop marking it.

> ** `kbmemfree` shows the amount of free memory available in kilobytes.
> ** `kbmemused` shows the amount of used memory in kilobytes.
> ** `kbcommit` shows the amount of memory in kilobytes needed for the current workload, which estimates how much RAM or swap is required to guarantee that it doesn't run out of memory.

Reviewer: Make the following punctuation edits GLOBALLY for similar texts throughout; I will stop marking them.
```suggestion
** `kbmemfree`, which shows the amount of memory available, in kilobytes
** `kbmemused`, which shows the amount of used memory, in kilobytes
** `kbcommit`, which shows the amount of RAM or swap memory needed for the current workload, in kilobytes 
```

.Sar command memory stats
image::mruntime-tuning-sar-memory.png[Sar command Memory stats]

== Disk Resource

> Some Mule components log writing operations and scenarios with large payloads treated as repeatable streams exhibit, where the exhausted resource are I/O (input or output) requests. Monitor the disk resource using the Sar command `sar -d -p $INTERVAL $COUNT`.

Reviewer: The first sentence is unclear:
Are the Mule components logging two things, writing operations and scenarios with large payloads?
What does "treated as repeatable streams exhibit" modify: log, or payloads?
What does "where the exhausted resources are input or output requests" modify?
Please reword and run by a peer writer.

Aside from that:
```suggestion
Some Mule components log writing operations and scenarios with large payloads treated as repeatable streams exhibit, where the exhausted resources are input or output requests. Monitor the disk resource using the Sar command `sar -d -p $INTERVAL $COUNT`.
```

Main metrics to pay attention are:

> ** `await` shows the average time in milliseconds for I/O requests issued to serve to the device to, which includes the time spent by the requests in the queue and the time spent servicing them.

Reviewer: I could not figure out "I/O requests issued to serve to the device to," so I attempted a rewording:
```suggestion
** `await`, which shows the average time, in milliseconds, that I/O requests are processed by the device, including the time requests spend waiting in the queue before they are processed.
```

> ** `%util` helps to understand the saturation of the device showing the percentage of CPU time during which I/O requests are issued to bandwidth utilization for the device.

Reviewer: I cannot understand "time during which I/O requests are issued to bandwidth utilization for the device." Please show this to a peer and reword for clarity.
Otherwise,
```suggestion
** `%util`, which indicates device saturation by showing the percentage of CPU time during which I/O requests are issued to bandwidth utilization for the device.
```

.Sar command disk stats
image::mruntime-tuning-sar-disk.png[Sar command Disk stats]

== Network Resource

> One of the main purposes of Mule is to interconnect with several services that communicate by the net. Because of this, the network infrastructure can become a bottleneck and saturated. You need to know the bandwidth and monitor the utilization, a simple way to do it is using the Sar command `sar -n DEV $INTERVAL $COUNT`.

Reviewer: ```suggestion
Because Mule connects several services communicating through the internet, network use can become saturated and stall. To manage this, you need to know the available bandwidth and monitor the utilization. A simple way to do this is by using the Sar command `sar -n DEV $INTERVAL $COUNT`.
```

Main metrics to pay attention are:

** `rxpck/s` shows the total number of packets received per second.

> ** `txpck/s` shows the total number of packets transmitted per second.

Reviewer: ditto prev. GLOBAL edits

.Sar command network stats
image::mruntime-tuning-sar-network.png[Sar command Network stats]

== CloudHub Monitoring

> To monitor Mule apps performance running in a CloudHub environment, you can use:

Reviewer: ```suggestion
You can use either the Anypoint Runtime Manager dashboard or Anypoint Monitoring to monitor Mule app performance in a CloudHub environment:
```

> * Runtime Manager Dashboard +

Reviewer: ```suggestion
* Runtime Manager dashboard +
```

> The tool allows you to visualize key performance metrics such as messages or transactions, CPU and memory usage, and specific time period metrics such as last hour, last 24 hours and last week.

Reviewer: ```suggestion
This tool enables you to visualize key performance metrics such as numbers of messages or transactions, CPU and memory use, and metrics during specific time periods, such as the past hour, the past 24 hours, or the past week.
```

.Runtime Manager Dashboard
image::mruntime-tuning-runtimemanager.png[Runtime Manager Dashboard]

* Anypoint Monitoring +

> The tool enables you to keep track and visualize your performance metrics by using built-in or custom dashboards exposing the metrics of your applications through visual representations of your resources. Additionally, the tool enables you to detect anomalies, troubleshoot issues, and see trends differently from the raw form.

Reviewer: ```suggestion
This tool enables you to track and visualize performance metrics by using built-in or custom dashboards showing visual representations of your resources. Additionally, Anypoint Monitoring enables you to detect anomalies, troubleshoot issues, and see trends that you might not notice by looking at raw data.
```
